fzakaria/slf4j-timbre#54, fzakaria/slf4j-timbre#45: Support SLF4j 2.x, timbre 6.x

chrisbetz opened the pull request:
Hi Farid (@fzakaria),

happy to use slf4j for quite some while now. Now I'm ready to update my projects to SLF4J 2.x and timbre 6.x, so I had to create this PR first:

Replaces the SLF4J 1.7.x Binders with an SLF4J 2.x ServiceProvider. Works in example ("integration-test").

I'd be happy to see this, would simplify my life :) Anyway: Thank you!!

Commit:
fzakaria/slf4j-timbre#54, fzakaria/slf4j-timbre#45: Update to work with SLF4j 2.x, timbre 6.x

Replaces the SLF4J 1.7.x Binders with an SLF4J 2.x ServiceProvider. Works in example ("integration-test").

rome-user commented on integration_tests/timbre5/project.clj:
```diff
@@ -1,8 +1,8 @@
-  :aot [example.core])
\ No newline at end of file
+            :main example.core
+            :aot [example.core])
```
Any particular reason for changing the indentation?

chrisbetz replied:
No, auto-format in Cursive and "don't care for whitespace"-setting in git client. No particular reason besides... it's just whitespace. Sorry.

rome-user commented on integration_tests/timbre5/src/example/core.clj:
```diff
@@ -15,4 +15,8 @@
+
+
+#_;
+        (-main)
```
Is the call to `-main` necessary for the test?

chrisbetz replied:
No, it just simplifies testing from a REPL. That's why I made the reader skip (-main) using the "#_" reader macro. The comment is just to ease readability in the editor.

Nothing essential, take it, leave it, it does not matter for functionality.

PavlosMelissinos replied:
If you'd rather keep it, you could put it in a rich comment. Makes the intent clearer, compared to that reader macro.

rome-user commented on project.clj:
```diff
@@ -2,29 +2,27 @@
-              :sub ["integration_tests/timbre5"]
-              :aliases {"run-integration-tests" ["sub" "do" "clean," "deps,"
-                                                 "update-in" ":dependencies" "conj" "[com.fzakaria/slf4j-timbre \"lein-git-inject/version\"]" "--" "run"]}}}
+              :plugins      [[lein-midje "3.2.2"]
+                             [lein-sub "0.3.0"]
+                             [day8/lein-git-inject "0.0.15"]]
+              :sub          ["integration_tests/timbre5"]
+              :aliases      {"run-integration-tests" ["sub" "do" "clean," "deps,"
+                                                      "update-in" ":dependencies" "conj" "[com.fzakaria/slf4j-timbre \"lein-git-inject/version\"]" "--" "run"]}}}
```
I'm assuming nothing is being changed here, but it's hard to tell.

chrisbetz replied:
No, just whitespaces. Sorry, again: It's my "don't care" setting.

rome-user commented on src/slf4j_timbre/timbre_service_provider.clj:
```diff
@@ -0,0 +1,40 @@
+           (org.slf4j.spi MDCAdapter))
+  (:gen-class :name com.github.fzakaria.slf4j.timbre.TimbreServiceProvider
+              :implements [org.slf4j.spi.SLF4JServiceProvider]
+              #_#_:methods [[getLoggerFactory [] org.slf4j.ILoggerFactory]
+                        [getMarkerFactory [] org.slf4j.IMarkerFactory]
+                        [getMDCAdapter [] org.slf4j.spi.MDCAdapter]
+                        [getRequestedApiVersion [] String]
+                        [initialize [] void]]
+              )
+  )
```
This should probably be removed since it's all commented out.

chrisbetz replied:
Yeah, right.

rome-user commented on test/slf4j_timbre/t_adapter.clj:
```diff
@@ -17,65 +18,68 @@
-     "one two three four" nil                    (Exception. "test")
-     "one two {} {}"      ["three" "four"]       (Exception. "test"))))
\ No newline at end of file
+                      (tabular
+                        (facts
+                          (timbre/with-context {:foo "preserved"}
+                                               (invoke-each logger ?args)
+                                               (invoke-each logger marker ?args)) => anything ; for side effects only
```
It's hard to tell if you are just re-formatting code or making changes here.

rome-user reviewed (commented):
I'm really happy to see that you've taken effort in updating bindings for my favorite logging library! With that said, I do have a few comments.

Commit:
Cleanup, see fzakaria/slf4j-timbre#61

My last commit introduced several whitespace changes, making it harder to spot the real differences. I cleaned these to simplify comparison. Also removed commented gen-class and obsolete requires in `slf4j-timbre.timbre-service-provider` namespace.

Commit:
Cleanup, see fzakaria/slf4j-timbre#61

My last commit introduced several whitespace changes, making it harder to spot the real differences. I cleaned these to simplify comparison. Also removed commented gen-class and obsolete requires in `slf4j-timbre.timbre-service-provider` namespace. (second try)

Commit:
Cleanup, see fzakaria/slf4j-timbre#61

My last commit introduced several whitespace changes, making it harder to spot the real differences. I cleaned these to simplify comparison. Also removed commented gen-class and obsolete requires in `slf4j-timbre.timbre-service-provider` namespace. (third try)

chrisbetz commented:
> I'm really happy to see that you've taken effort in updating bindings for my favorite logging library! With that said, I do have a few comments.

I removed all the re-formatting, so hopefully, the PR is now easier to check. Sorry for that, it's just my "I don't care for whitespaces" attitude ;) Didn't want to complicate things on your end, though. The actual PR is really small ;) (ah, and don't ask for the magic "2.0.99" number, I copied that from the SLF4J source and it worked.

rome-user reviewed (approved):
Your PR looks very clean now! :D 
Let's hope @fzakaria can take a look soon

fzakaria commented:
@rufoa has been the maintainer for a very long time.  🙏

rufoa commented:
Thank you @chrisbetz for your work on this! And to @PavlosMelissinos @rome-user for your reviews.

It's not clear to me that it will be possible to retain support for slf4j 1.7 emitters, or how much of a problem this will be in practice. I suppose it's more important that we add 2.x support, as your kind PR does, than be paralysed forever by this potential issue.

(And thanks for the timbre 6 support, which I had also intended to add.)

vincentjames501 commented:
@rufoa any thoughts on merging this? If you're worried about backwards compatibility maybe this is enough of an upgrade to warrant a 1.x.x release to indicate a potentially breaking change?

devurandom commented:
@rufoa, @rome-user: Do you have objections left that prevent merging this? It seems many libraries are updating to SLF4J 2 and to allow slf4j-timbre to be used in that context, it needs to be updated. I supposed slf4j-timbre was stable and mature enough that you can just make a cut, move to version 0.4 and stop supporting SLF4J 1.7? That way, if any bug fix would be necessary for older applications using SLF4J 1.7, and you want to put in the effort to support those, you could release another 0.3 version.

Merging this would close most issues / PRs in this repository:

Closes: https://github.com/fzakaria/slf4j-timbre/issues/42
Closes: https://github.com/fzakaria/slf4j-timbre/issues/45
Closes: https://github.com/fzakaria/slf4j-timbre/issues/54
Closes: https://github.com/fzakaria/slf4j-timbre/pull/59
Closes: https://github.com/fzakaria/slf4j-timbre/pull/60
Closes: https://github.com/fzakaria/slf4j-timbre/pull/62
Closes: https://github.com/fzakaria/slf4j-timbre/pull/63

rome-user commented:
No objections. I've approved the changes a while ago, but I am not maintainer of the library.

I have personally moved on to Logback and haven't used Timbre in a while now

rufoa commented:
working on this now!

rufoa commented:
Please could you all try this out:
https://clojars.org/com.fzakaria/slf4j-timbre/versions/0.3.21-10-299a94e-SNAPSHOT

Thank you!

@devurandom @rome-user @vincentjames501 @chrisbetz

dekelpilli commented:
@rufoa It seems that there are a few reflection issues here - https://github.com/fzakaria/slf4j-timbre/blob/slf4j2/src/slf4j_timbre/service_provider.clj - all `this` labels need to be type-hinted (as ^com.github.fzakaria.slf4j.timbre.TimbreServiceProvider)

rufoa commented:
Thanks @dekelpilli, fixed in https://clojars.org/com.fzakaria/slf4j-timbre/versions/0.3.21-11-cb46216-SNAPSHOT

vincentjames501 commented:
@rufoa , any plans to deploy this?

rufoa commented:
yes! will get on it

rufoa commented:
Released 0.4.0

https://clojars.org/com.fzakaria/slf4j-timbre/versions/0.4.0

Thanks everyone!
